File: bats.ml
(******************************************************************************************************
  Course    : Programming Languages 1 - Assignment 2 - Exercise 1
  Authors   : Nikolaos Papadis (nikpapadis@gmail.com), Christos Markou (chrs.markx86@gmail.com)
  Date      : August 27, 2014
  Note      : We solve the problem in three steps:
              1) After we have parsed the input, we create the map.
              2) We create all valid lines (as edges of a graph, whose verteces are the animals), 
              using the Bresenham supercover algorithm.
              3) We find the shortest path using Dijkstra's Shortest Path Algorithm. 
                 In the end, we print the shortest path's length with precision of two decimal points.
  -----------
  School of ECE, National Technical University of Athens.
******************************************************************************************************)

open Printf
open Str

(* Initialisations *)

let n = ref 0
let m = ref 0
let k = ref 0
let numOfAnimals = ref 0
let target = ref 0
let myMap = ref (Array.make_matrix 0 0 "~");;


(* Implementation of Dijkstra's Algorithm in OCaml taken from http://rosettacode.org/wiki/Dijkstra%27s_algorithm#OCaml *)

type vertex = int
type weight = float
type neighbor = vertex * weight
module VertexSet = Set.Make(struct type t = weight * vertex let compare = compare end)
 
let dijkstra (src:vertex) (adj_list:neighbor list array) : weight array * vertex array =
  let n = Array.length adj_list in
  let min_distance = Array.make n infinity in
  min_distance.(src) <- 0.;
  let previous = Array.make n (-1) in
  let rec aux vertex_queue =
    if not (VertexSet.is_empty vertex_queue) then
      let dist, u = VertexSet.min_elt vertex_queue in
      let vertex_queue' = VertexSet.remove (dist, u) vertex_queue in
      let edges = adj_list.(u) in
      let f vertex_queue (v, weight) =
        let dist_thru_u = dist +. weight in
        if dist_thru_u >= min_distance.(v) then
          vertex_queue
        else begin
          let vertex_queue' = VertexSet.remove (min_distance.(v), v) vertex_queue in
          min_distance.(v) <- dist_thru_u;
          previous.(v) <- u;
          VertexSet.add (min_distance.(v), v) vertex_queue'
        end
      in
      aux (List.fold_left f vertex_queue' edges)
  in
  aux (VertexSet.singleton (min_distance.(src), src));
  min_distance, previous
 
let shortest_path_to (target : vertex) (previous : vertex array) : vertex list =
  let rec aux target acc =
    if target = -1 then
      acc
    else
      aux previous.(target) (target :: acc)
  in
  aux target []

(* End of Dijkstra *)


(* File parsing function *)

let parseFile =
    let filename = Sys.argv.(1) in
    let example myList = Str.split (regexp "[ \n]+") myList in
    let first ((a:string)::(b:string)::(c:string)::[]) = a in
    let second ((a:string)::(b:string)::(c:string)::[]) = b in
    let third ((a:string)::(b:string)::(c:string)::[]) = c in
    let lines = ref [] in   
    let chan = open_in filename in
    let myLine = input_line chan in
    let myLineSplitted = example myLine in
      n := int_of_string(first(myLineSplitted));
      m := int_of_string(second(myLineSplitted));
      k := int_of_string(third(myLineSplitted));
      
    let tempArray = (Array.make_matrix !n !m "~") in 
        myMap := Array.append !myMap tempArray;
        let id = ref 0 in    
    try
        while true; do
            let myLine = input_line chan in 
            let [i;j;ch] = (example myLine) in
            let x = int_of_string(i) and y = int_of_string(j) in
            (!myMap).(x).(y) <- ch;
            
            if ch <> "-" then 
                lines := (!id,x,y,ch) :: (!lines);
            if ch = "A" then target:= !id;
            if ch <> "-" then id := (!id) +1;
        done; []
    with End_of_file ->
      close_in chan;
      numOfAnimals := List.length(!lines);
      numOfAnimals := (!id);
      (List.rev !lines);;


(* Bresenham supercover algorithm taken from http://lifc.univ-fcomte.fr/home/~ededu/projects/bresenham/ and implemented in OCaml by us *) 

let bresenham ((x1:int), (y1:int), (x2:int), (y2:int)) : bool = 
    let dx = ref 0 in
        let dy = ref 0 in
        let ystep = ref 0 in
        let xstep = ref 0 in
        let ddy = ref 0 in
        let ddx = ref 0 in
        let error = ref 0 in
        let result = ref true in
        let x = ref 0 in
        let y = ref 0 in

        let errorprev = ref 0 in
        
        x:=x1;
        y:=y1;
        dx := x2-x1;
        dy := y2-y1;
        
        if !dy<0 then 
        begin
            ystep := -1; 
            dy := -(!dy);
        end
        else begin 
            ystep:=1;
        end;

        if (!dx<0) then 
        begin
            xstep := -1; 
            dx := -(!dx);
        end
        else begin 
            xstep:=1;
        end;
        ddy := 2* (!dy);
        ddx := 2* (!dx);  
        if (!ddx) >= (!ddy) then
        begin 
            errorprev := !dx;
            error := !dx;
            let i = ref 0 in
            while ((!i) < (!dx)) && ((!result)=true) do
                x:= (!x)+(!xstep);
                error:= (!error) + (!ddy);
                if ((!error)>(!ddx)) then
                    begin
                        y:= (!y)+(!ystep);
                        error:= (!error) - (!ddx);

                        if(((!error)+(!errorprev)) < (!ddx)) then
                        begin
                            let tempx = (!x) in
                            let tempy = ((!y)-(!ystep)) in
                            if ((!myMap).(tempx).(tempy) = "-") then
                            begin
                                if (!result) then result := false;
                            end;
                        end

                        else if(((!error)+(!errorprev)) > (!ddx)) then
                        begin
                            let tempx = ((!x) - (!xstep))in
                            let tempy = (!y) in
                            if ((!myMap).(tempx).(tempy) = "-") then
                            begin
                                if (!result) then result := false;
                            end;
                        end

                        else 
                        begin
                            let tempx = ((!x))in
                            let tempy = ((!y)-(!ystep)) in
                            if ((!myMap).(tempx).(tempy) = "-") then
                            begin
                                if (!result) then result := false;
                            end;
                            let tempx = ((!x) - (!xstep)) in
                            let tempy = (!y) in
                            if ((!myMap).(tempx).(tempy) = "-") then
                            begin
                                if (!result) then result := false;
                            end;

                        end;
                    end;
                let tempx = (!x) in
                let tempy = (!y) in
                if ((!myMap).(tempx).(tempy) = "-") then
                begin
                    if (!result) then result := false;
                end;
                if (!result) then errorprev := (!error); 
                i := (!i) + 1;   
            done;
        end
        else
        begin
        
            if (!result) then 
            begin
                errorprev := !dy;
                error := !dy;
                let i = ref 0 in
                while ((!i) < (!dy)) && ((!result)=true) do
                    y:= (!y) + (!ystep);
                    error := (!error) + (!ddx);
                    if ((!error) > (!ddy)) then
                        begin
                            x := (!x) + (!xstep);
                            error := (!error) - (!ddy);

                            if(((!error)+(!errorprev)) < (!ddy)) then
                            begin
                                let tempx = ((!x) - (!xstep)) in
                                let tempy = (!y) in

                                if ((!myMap).(tempx).(tempy) = "-") then
                                begin
                                    if (!result) then result := false;
                                end;
                            end

                            else if(((!error)+(!errorprev)) > (!ddy)) then
                            begin
                                let tempx = (!x) in
                                let tempy = ((!y)-(!ystep)) in
                                if ((!myMap).(tempx).(tempy) = "-") then
                                begin
                                    if (!result) then result := false;
                                end;
                            end

                            else 
                            begin
                                
                                let tempx = ((!x)- (!xstep)) in
                                let tempy = (!y) in
                                if ((!myMap).(tempx).(tempy) = "-") then
                                begin
                                    if (!result) then result := false;
                                end;
                                let tempx = ((!x)) in
                                let tempy = ((!y)-(!ystep)) in
                                if ((!myMap).(tempx).(tempy) = "-") then
                                begin
                                    if (!result) then result := false;
                                end;

                            end;
                        end;
                    let tempx = (!x) in
                    let tempy = (!y) in
                    if ((!myMap).(tempx).(tempy) = "-") then
                    begin
                        if (!result) then result := false;
                    end;
                    if (!result) then errorprev := (!error); 
                    i := (!i) + 1;   
                done;
            end;
        end;
        (!result);;


(* Function that calculates the distance between two points of the plane *)

let calcDistance ((x1,y1):(int*int)) ((x2,y2):(int*int)) = 
    let dx = abs(x1-x2) in
    let dy = abs(y1-y2) in
      (sqrt(float_of_int(dx*dx + dy*dy)) : weight);;


(* Definition of the graph that has the animals as verteces and the valid lines as edges*)

let myG = ref (Array.make !numOfAnimals ([]:neighbor list));;


(* Recusive function that creates all edges of the graph *)

let rec create_all_edges = function
    [] -> ();
  | ((id1,i1,j1,ch1)::ans) ->
        let tempAdj = (ref [] : neighbor list ref) in
        let rec find_my_adjacents = function
            [] -> ();
          | (id2,i2,j2,ch2)::ans ->
                let iAmAdj = bresenham (i1,j1,i2,j2) in
                
                let dist = calcDistance (i1,j1) (i2,j2) in 
                tempAdj := (!myG).(id1);  
                if iAmAdj then tempAdj := (id2,dist)::(!tempAdj);
                if iAmAdj then (!myG).(id1) <- (!tempAdj);
                tempAdj := (!myG).(id2);
                if iAmAdj then tempAdj := (id1,dist)::(!tempAdj);
                if iAmAdj then (!myG).(id2) <- (!tempAdj);           
                    find_my_adjacents ans;  
        in
            find_my_adjacents ans;
            create_all_edges ans;;


(* Main body of the program *)

let () =
    let animals = parseFile in 
    let _ = create_all_edges animals in
    let adj_list =(!myG) in 
    let toGo = !target in   
    let min_distance, previous = dijkstra 0 (adj_list) in
    printf "%0.*f\n" 2 min_distance.(toGo);

File: bats_java_full_version/BatsSubmition/myBatsSubmition/Bats.java
/******************************************************************************************************
  Course    : Programming Languages 1 - Assignment 2 - Exercise 1
  Authors   : Nikolaos Papadis (nikpapadis@gmail.com), Christos Markou (chrs.markx86@gmail.com)
  Date      : August 30, 2014
  Note      : We solve the problem in three steps:
              1) We create the map.
              2) We create all valid lines (as edges of a graph, whose verteces are the animals), 
              using the Bresenham supercover algorithm.
              3) We find the shortest path using Dijkstra's Shortest Path Algorithm. 
                 In the end, we print the shortest path's length with precision of two decimal points.
  -----------
  School of ECE, National Technical University of Athens.
******************************************************************************************************/

//package myButs;

import java.io.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.lang.*;

/**
 * @author Nick Papadis and chris
 * 
 */
public class Bats {

	/**
	 * @param args
	 */
	static char[][] building;
	static int[][] helpBuilding;
	static double targetX;
	static double targetY;
	public static void main(String[] args) {
		try {

			FileInputStream fstream = new FileInputStream(
					args[0].trim());
			DataInputStream in = new DataInputStream(fstream);
			BufferedReader br = new BufferedReader(new InputStreamReader(in));

			// here we do the reading of the input file!
			String line = br.readLine();
			String[] a = line.split(" ");
			int n = Integer.parseInt(a[0]);
			int m = Integer.parseInt(a[1]);
			int k = Integer.parseInt(a[2]);
			building = new char[n][m];
			helpBuilding = new int[n][m];
			int target=0;
			

			System.out.println();
			// create the whole table filling it with '~'
			for (int i = 0; i < n; i++)
				for (int j = 0; j < m; j++){
					building[i][j] = '~';
					helpBuilding[i][j]=-1;
				}

			ArrayList<interestingPoint> Animals = new ArrayList<interestingPoint>();
			ArrayList<interestingPoint> Walls = new ArrayList<interestingPoint>();

			for (int i = 0; i < k; i++) {
				line = br.readLine();
				String[] temp = line.split(" ");
				int tempI = Integer.parseInt(temp[0]);
				int tempJ = Integer.parseInt(temp[1]);
				char Inside = temp[2].charAt(0);
				

				// store in two lists the animals and the walls!!!
				if (Inside == 'B') {
					interestingPoint e = new interestingPoint(tempI + 0.5,
							tempJ + 0.5, 'B');
					building[tempI][tempJ] = 'B';
					Animals.add(e);
				} else if (Inside == 'A') {
					interestingPoint e = new interestingPoint(tempI + 0.5,
							tempJ + 0.5, 'A');
					Animals.add(e);
					targetX = tempI + 0.5;
					targetY = tempJ + 0.5;
					building[tempI][tempJ] = 'A';
				} else {
					interestingPoint e = new interestingPoint(tempI + 0.5,
							tempJ + 0.5, 'w');
					Walls.add(e);
					building[tempI][tempJ] = 'w';
				}
			}
			// never forget to close the stream!
			in.close();

			// now that we have finished with the reading we are ready to proceed

			// now we are ready to create the lines of sight!!!
			ArrayList<ArrayList> lines_of_sight = new ArrayList<ArrayList>();
			
			for (int i = 0; i < Animals.size(); i++) {
				for(int j=i+1; j < Animals.size(); j++){
				 double StartX = Animals.get(i).getX();
				 double StartY = Animals.get(i).getY();
				 double EndX = Animals.get(j).getX();
				 double EndY = Animals.get(j).getY();
				ArrayList<simplePoint> e = new ArrayList<simplePoint>();
				e = Bresenham_supercover(StartX, StartY,EndX, EndY);
				
				if (!e.isEmpty()) lines_of_sight.add(e); // add only the non empty lines
				}
			}
			
			
			// now you need the clear lines!!!
			ArrayList<ArrayList> clear_lines = new ArrayList<ArrayList>();
			clear_lines = lines_of_sight;
			
			
			// now that you have the clear lines you need to create the edges of the graph.
			ArrayList<Edge> edges = new ArrayList<Edge>();
			ArrayList<Vertex> nodes = new ArrayList<Vertex>();
			int counterNodes=0;
			int [] times = new int [clear_lines.size()];
			for (int i = 0; i < clear_lines.size(); i++) times[i]=0;
			
			
			for (int i = 0; i < clear_lines.size(); i++){
				ArrayList<simplePoint> tempList = new ArrayList<simplePoint>();
				tempList = clear_lines.get(i);
				
				int size = tempList.size();
								
				if (size!=1){
					double x1 = tempList.get(0).getX(); 
					double y1 = tempList.get(0).getY(); 
					double x2 = tempList.get(size-1).getX(); 
					double y2 = tempList.get(size-1).getY();
					int tempX=(int) (x1-0.5);
					int tempY=(int) (y1-0.5);
					int tempX2=(int) (x2-0.5);
				    int tempY2=(int) (y2-0.5);
					
					int temp=0;
					
					if(helpBuilding[tempX][tempY]!=-1){ 
						temp=helpBuilding[tempX][tempY];
						}
					else temp=counterNodes;
										
					Vertex mySource= new Vertex("VertexNo"+temp,"VertexNo"+temp,x1,y1);

					if (building[tempX][tempY]!='d'){
						building[tempX][tempY]='d';
						int index = helpBuilding[tempX][tempY];
						if (index==-1){
							helpBuilding[tempX][tempY]=counterNodes;}
					
						
						nodes.add(mySource);
						counterNodes++;
						}					
					
					int temp2=0;
					if(helpBuilding[tempX2][tempY2]!=-1) { 
						temp2=helpBuilding[tempX2][tempY2];
					}
					else temp2=counterNodes;
					
					Vertex myDest= new Vertex("VertexNo"+temp2,"VertexNo"+temp2,x2,y2);
										
				    if (building[tempX2][tempY2]!='d'){
				    	building[tempX2][tempY2]='d';
						int index = helpBuilding[tempX2][tempY2];
						if (index==-1){helpBuilding[tempX2][tempY2]=counterNodes;}

						nodes.add(myDest);
						counterNodes++;
					}

					double dist = Math.sqrt( Math.pow( (x1-x2), 2)+Math.pow(  (y1-y2), 2) );
					Edge tempEdge = new Edge("edgeNo"+i,mySource, myDest,dist);
					
					edges.add(tempEdge);				
				}
			}
			
			int myX = (int)(targetX-0.5);
			int myY = (int)(targetY-0.5);
			
			target = helpBuilding[myX][myY];
			
			
			//now create the appropriate graph!!!
					
			Graph graph = new Graph(nodes, edges);
		    DijkstraAlgorithm dijkstra = new DijkstraAlgorithm(graph);

		    dijkstra.execute(nodes.get(0));

		    double tempResult = dijkstra.getShortestDistance(nodes.get(target));
		    LinkedList<Vertex> path = dijkstra.getPath(nodes.get(target));
		    
			
			//implement Dijkstra on your graph to find the path!
			double finalResult = Math.round( tempResult*100.0)/100.0;
			System.out.println(finalResult);
			
			/* insert the main code here !!!! */

		} 
		catch (Exception e) {
			e.printStackTrace();
		}
	}

	/*
	 * algorithm from:
	 * http://lifc.univ-fcomte.fr/home/~ededu/projects/bresenham/
	 */
	public static ArrayList Bresenham_supercover(double x1, double y1,double x2, double y2) {
		ArrayList<simplePoint> points = new ArrayList<simplePoint>();
		ArrayList<simplePoint> empty = new ArrayList<simplePoint>();

		double i; // loop counter
		double ystep, xstep; // the step on y and x axis
		double error; // the error accumulated during the increment
		double errorprev; // *vision the previous value of the error variable
		double y = y1, x = x1; // the line points
		double ddy, ddx; // compulsory variables: the double values of dy and dx
		double dx = x2 - x1;
		double dy = y2 - y1;
		simplePoint temPoint = new simplePoint(x1, y1); // first point
		points.add(temPoint);
		// here check if the point is a wall!!!
		int myX = (int)(x1-0.5);
		int myY = (int)(y1-0.5);
		
		if(building[myX][myY]=='w'){
			return empty;
		}
		// NB the last point can't be here, because of its previous point (which
		// has to be verified)
		if (dy < 0) {
			ystep = -1;
			dy = -dy;
		} else
			ystep = 1;
		if (dx < 0) {
			xstep = -1;
			dx = -dx;
		} else
			xstep = 1;
		ddy = 2 * dy; // work with double values for full precision
		ddx = 2 * dx;
		if (ddx >= ddy) { // first octant (0 <= slope <= 1)
			// compulsory initialization (even for errorprev, needed when
			// dx==dy)
			errorprev = error = dx; // start in the middle of the square
			for (i = 0; i < dx; i++) { // do not use the first point (already
										// done)
				x += xstep;
				error += ddy;
				if (error > ddx) { // increment y if AFTER the middle ( > )
					y += ystep;
					error -= ddx;
					
					// three cases (octant == right->right-top for directions
					// below):
					if (error + errorprev < ddx) { // bottom square also
						simplePoint temPoint2 = new simplePoint( x , y-ystep); 											
						points.add(temPoint2);
						 myX = (int)(x-0.5);
						 myY = (int)(y-ystep-0.5);
						
						if(building[myX][myY]=='w'){
							return empty;
						}

					} else if (error + errorprev > ddx) { // left square also
						simplePoint temPoint2 = new simplePoint(x - xstep, y); 												
						points.add(temPoint2);
						myX = (int)(x- xstep-0.5);
						myY = (int)(y-0.5);
						
						if(building[myX][myY]=='w'){
							return empty;
						}
					} else { // corner: bottom and left squares also
						simplePoint temPoint2 = new simplePoint( x , y-ystep); 												
						points.add(temPoint2);
						myX = (int)(x-0.5);
						myY = (int)(y-ystep-0.5);
						
						if(building[myX][myY]=='w'){
							return empty;
						}
						simplePoint temPoint3 = new simplePoint(x - xstep, y); 															
						points.add(temPoint3);
						myX = (int)(x-xstep-0.5);
						myY = (int)(y-0.5);
						
						if(building[myX][myY]=='w'){
							return empty;
						}
					}
				}
				simplePoint temPoint4 = new simplePoint(x, y); 
				points.add(temPoint4);
				myX = (int)(x-0.5);
				myY = (int)(y-0.5);
				
				if(building[myX][myY]=='w'){
					return empty;
				}
				errorprev = error;
			}
		} else { // the same as above
			errorprev = error = dy;
			for (i = 0; i < dy; i++) {
				y += ystep;
				error += ddx;
				if (error > ddy) {
					x += xstep;
					error -= ddy;
					
					if (error + errorprev < ddy) {
						simplePoint temPoint2 = new simplePoint(x - xstep, y); 													
						points.add(temPoint2);
						myX = (int)(x- xstep-0.5);
						myY = (int)(y-0.5);
						
						if(building[myX][myY]=='w'){
							return empty;
						}
					} else if (error + errorprev > ddy) {
						simplePoint temPoint2 = new simplePoint(x,y - ystep); 													
						points.add(temPoint2);
						myX = (int)(x-0.5);
						myY = (int)(y-ystep-0.5);
						
						if(building[myX][myY]=='w'){
							return empty;
						}
					} else {
						simplePoint temPoint2 = new simplePoint(x - xstep,y);
						points.add(temPoint2);
						myX = (int)(x-xstep-0.5);
						myY = (int)(y-0.5);
						
						if(building[myX][myY]=='w'){
							return empty;
						}
						simplePoint temPoint3 = new simplePoint(x, y - ystep); 																			
						points.add(temPoint3);
						myX = (int)(x-0.5);
						myY = (int)(y-ystep-0.5);
						
						if(building[myX][myY]=='w'){
							return empty;
						}
					}
				}
				simplePoint temPoint4 = new simplePoint(x, y); // first point
				points.add(temPoint4);
				myX = (int)(x-0.5);
				myY = (int)(y-0.5);
				
				if(building[myX][myY]=='w'){
					return empty;
				}
				errorprev = error;
			}
		}
	
		return points;
	}
	
}